GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs

```diff
@@ -0,0 +1,179 @@
+        /// on Unity's main thread, therefore the action should only be performing operations that can be run off the
+        /// main thread. 
+        /// </param>
+        public OAuth2Server(Action<KeyValuePair<string, string>> onResponseAction)
+        {
+            _onResponseAction = onResponseAction;
+            _callbackEndpoint = string.Format("http://localhost:{0}/{1}/", ServerPort, Path.GetRandomFileName());
+            _httpListener = new HttpListener();
+            _httpListener.Prefixes.Add(_callbackEndpoint);
+        }
+
+        /// <summary>
+        /// The callback endpoint on which the server is listening.
+        /// </summary>
+        public string CallbackEndpoint
+        {
+            get { return _callbackEndpoint; }
+        }
+
+        /// <summary>
+        /// Allow this instance to start listening for incoming requests containing authorization code or error data
+        /// from OAuth2. Server will stop after processing the first request.
+        /// </summary>
+        public void Start()
+        {
+            _httpListener.Start();
+            // Server will only respond to the first request, therefore one new thread should handle it just fine.
+            new Thread(() => { ProcessContext(_httpListener.GetContext()); }).Start();
+        }
+
+        /// <summary>
+        /// Processes the object as an HttpListenerContext instance and retrieves authorization response. Invokes the
+        /// response handler action on the response, responds to request with a string asking the user to close tab,
+        /// and stops the server from listening for future incoming requests.
+        /// </summary>
+        private void ProcessContext(HttpListenerContext context)
+        {
+            var authorizationResponse = GetAuthorizationResponse(context.Request.Url);
+            _onResponseAction(authorizationResponse);
+            context.Response.KeepAlive = false;
+            var responsebBytes = Encoding.UTF8.GetBytes(CloseTabText);
+            var outputStream = context.Response.OutputStream;
+            outputStream.Write(responsebBytes, 0, responsebBytes.Length);
+            outputStream.Flush();
+            outputStream.Close();
+            context.Response.Close();
+            Stop();
+        }
+
+        /// <summary>
+        /// Returns a key value pair corresponding to the authorization response sent from OAuth2 authorization page.
+        /// Logs error message and throws ArgumentException if the uri contains invalid params.
+        /// </summary>
+        /// <param name="uri">The uri of the incoming request</param>
+        /// <exception cref="ArgumentException">Exception thrown if the uri contains invalid params.</exception>
+        internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
+        {
+            if (!UriContainsValidQueryParams(uri))
```

Comment:
To be honest, I was a bit confused about what exactly the code was trying to do until the latest PR that simplifies things a bit. The current approach relies on calling GetQueryParamsFromUri() which is a bit inefficient. 

So I'm going to suggest something that is hopefully simpler and that provides an error message that specifically identifies the issue as opposed to a generic message that "Uri query is not valid".

```
internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
{
  var query = GetQueryString(uri);
  return GetCodeOrErrorResponsePair(query);
}

internal static string GetQueryString(uri)
{
var query = uri.Query;
if (string.IsNullOrEmpty(query))
{
  throw new ArgumentException("URI is missing query parameters");
}
if (query[0] != "?")
{
  throw new ArgumentException("Expect first character of URI query is \"?\"");
}
return query.Substring(1);
}

internal static KeyValuePair<string, string> GetCodeOrErrorResponsePair(string query)
{
// The authorization response should consist of a single query parameter.
var paramPairs = query.Split('&');
if (paramPairs.Length != 1)
{
  throw new ArgumentException(string.Format("Unexpected number of URI parameters {0}", paramPairs.Length));
}

// The single query parameter should be an equals-separated key/value pair.
var keyAndValue = paramPairs[0].Split('=');
if (keyAndValue.Length != 2)
{
  throw new ArgumentException(string.Format("URI parameter pair has {0} components", keyAndValue.Length));
}
// The key for the single query parameter should either be "code" or "error".
var key = keyAndValue[0];
var value = Uri.UnescapeDataString(keyAndValue[1];
if (key != "code" || key != "error")
{
  throw new ArgumentException(string.Format("Unexpected URI parameter key \"{0}\"", key));
}
return new KeyValuePair<string, string>() {key, value}; // Not sure about format here, whatever is correct/concise.
}
```

Reply:
Resolved.